File: 0972-knight-dialer/0972-knight-dialer.cpp
#include <iostream>
#include <vector>

using namespace std;
class Solution {
public:
    int knightDialer(int n) {


        long long MODULO = 7 + 1e9 ;

        vector<vector<int>> moves = {
            {4, 6},    // 0
            {6, 8},    // 1
            {7, 9},    // 2
            {4, 8},    // 3
            {0, 3, 9}, // 4
            {},        // 5 (no valid knight moves to/from 5)
            {0, 1, 7}, // 6
            {2, 6},    // 7
            {1, 3},    // 8
            {2, 4}     // 9
        };

       
        vector<vector<int>> dp(n+1,vector<int>(10,0));

        cout << size(dp) << endl;

        for (int i = 0; i < 10 ; i++){
            dp[1][i] = 1;
        }

        // now we do the recursiveness

        for (int i = 2; i < n+1; i++){
            for (int digit = 0; digit < 10; digit++){
                for (int c: moves[digit]){

                    dp[i][digit] = (dp[i][digit] + dp[i-1][c] ) % MODULO;
                }

            }
        }
        int res = 0;
        for (int i = 0; i < 10; i ++){
            res = (res + dp[n][i]) % MODULO;
        }
        return res;
    }
};

File: 0972-knight-dialer/0972-knight-dialer.py
class Solution:
    def knightDialer(self, n: int) -> int:
        MODULO = pow(10,9) + 7
        
        moves = {
            0: [4, 6],
            1: [6, 8],
            2: [7, 9],
            3: [4, 8],
            4: [0, 3, 9],
            5: [],
            6: [0, 1, 7],
            7: [2, 6],
            8: [1, 3],
            9: [2, 4]
        }
        

        dp = [ [0 for _ in range(10)] for _ in range(n+1)]


        for i in range(10):
            dp[1][i] = 1
        

        for i in range(2, n+1):
            for digit in range(10):
                for move in moves[digit]:
                    dp[i][digit] = (dp[i][digit] + dp[i-1][move]) % MODULO  
        

        return sum(dp[n]) % MODULO  